### ISSUE TITLE
Feature request: Allow "anything" as a type in [pack]

### ISSUE THREAD

**Issue by mxa** [pure-data/pure-data#891]
This would allow for polymorphic/generic code and one could for example make an abstraction which could do what [swap] does but with symbols and floats mixed.

**Comment by umlaeute** [comment]
use `[zexy/pack]` instead :-)

zexy's `[pack]` implementation used to override the built-in `[pack]` to allow something like that, but people complained so loudly that i had to undo that.
either my design or the feature might be flawed (i'd rather bet on my design, but who knows...)

**Comment by porres** [comment]
Other similar externals are cyclone/pak and else/pack2, but they're more useful if you want all inputs to be hot.

But anyway, seems like an 'anything/a' in pack wouldn't be bad, and I wonder if they could also take lists, so you can join and merge two or more lists as well!

**Comment by porres** [comment]
> This would allow for polymorphic/generic code and one could for example make an abstraction which could do what [swap] does but with symbols and floats mixed.

why not also just allow swap to take symbols as well?

**Comment by porres** [comment]
> I wonder if they could also take lists, so you can join and merge two or more lists as well!

the problem with this is that you can't have a proper counterpart with unpack...

**Comment by mxa** [comment]
> > I wonder if they could also take lists, so you can join and merge two or more lists as well!
> 
> the problem with this is that you can't have a proper counterpart with unpack...

Not every time an unpack counterpart is needed.

**Comment by mxa** [comment]
> use `[zexy/pack]` instead :-)
> 
> zexy's `[pack]` implementation used to override the built-in `[pack]` to allow something like that, but people complained so loudly that i had to undo that.
> either my design or the feature might be flawed (i'd rather bet on my design, but who knows...)

ClaudiusMaximus mentioned `[lpack]` from pdlua as well.

**Comment by mxa** [comment]
> why not also just allow swap to take symbols as well?

honestly I don't understand why [swap] exists since its functionality can be built with more basic objects: `[pack f f]` and `[$2 $1(`

**Comment by porres** [comment]
many objects can be built with other basic ones. For instance, I can implement [moses] with [<] and [select], swap can also be implemented like this: 

<img width="252" alt="Screen Shot 2020-02-09 at 13 49 02" src="https://user-images.githubusercontent.com/5360142/74110636-09f27c80-4b43-11ea-976e-0a0632c60efa.png">

So yeah, it's just a matter of convenience.

**Comment by porres** [comment]
> Not every time an unpack counterpart is needed.

even though I like the idea of being able to join lists, I wouldn't know how to unjoin them with unpack, so this functionality could be better implemented into a new function in [list], actually I think there was a PR about it.

**Comment by porres** [comment]
>  I think there was a PR about it.

nope, there wasn't, there's a PR for [list join] and [list unjoin] but they do different things. Well, maybe we could open a new feature request or PR for that, but it'd be separate from this one. And btw, cyclone/join and cyclone/unjoin & else/merge and else/unmerge do that.

Now, there are two suggestions here: 
- a) make pack/unpack deal with either float or symbol; 
- b) make swap deal with either float or symbol

I could implement both suggestions into separate PRs, but not this month.

**Comment by umlaeute** [comment]
whatever you do, make sure that the implementation 
- is not performing much worse
- still allows for automatic distribution of list-messages over all the inlets

**Comment by porres** [comment]
as far as implementing a more flexible [swap], you can do it with a plain [list] object and don't need to add this functionality to [pack] so you can do that. 

<img width="391" alt="Screen Shot 2020-02-11 at 07 15 00" src="https://user-images.githubusercontent.com/5360142/74249944-c4e35d00-4c9e-11ea-8674-379d7b008b8c.png">

and where I am going with this is that we should have more examples of use cases for either expansion in functionality.

I don't have one right now for either, but ut feels to me that adding 'anything' to [pack] really looks useful

**Comment by porres** [comment]
hmm, how about list prepend for a more flexible [swap]? 

<img width="338" alt="Screen Shot 2020-02-11 at 07 21 45" src="https://user-images.githubusercontent.com/5360142/74250269-36231000-4c9f-11ea-95b8-0e23cb9f1f72.png">

**Comment by porres** [comment]
I deleted one of the messages above, anyway, as far as implementing a more flexible [swap], [list prepend] can do it. So we don't need to add this functionality to [pack] so you can do that. Using "$2 $1" in messages also swaps things as mentioned.

It still feels to me that adding 'anything' to [pack] looks useful though

**Comment by umlaeute** [comment]
keep in mind that `[list prepend]` is
- about 3-4 times slower than `[swap]`
- does not allow  for automatic distribution of list-messages over the inlets
